### PR TITLE
Check character boundaries before comparing string slices

### PR DESCRIPTION
**Pull request by djc**

### CODE REVIEW

**Review by valkum — COMMENTED** [review]
Wouldn't the line with `result.push_str(input.get(last_end..last_end + 1).unwrap());`  (533) also be susceptible to larger UTF-8 characters?

**Comment by djc** [comment]
@valkum good point! I decided to invest a little, and rewrote the `decode()` implementation with an explicit state machine that is a little shorter, quite a bit faster and more likely to be correct -- comes with benchmarks and tests.

File: instant-xml/src/impls.rs [hunk 1]
```diff
@@ -769,3 +770,23 @@ impl<'xml> FromXml<'xml> for IpAddr {
...
+        assert_eq!(decode("foo &lt; bar").unwrap(), "foo < bar");
+        assert_eq!(decode("foo &gt; bar").unwrap(), "foo > bar");
+        assert_eq!(decode("foo &quot; bar").unwrap(), "foo \" bar");
+        assert_eq!(decode("foo &apos; bar").unwrap(), "foo ' bar");
+        assert_eq!(decode("foo &amp;lt; bar").unwrap(), "foo &lt; bar");
+        assert_eq!(decode("&amp; foo").unwrap(), "& foo");
+        assert_eq!(decode("foo &amp;").unwrap(), "foo &");
+        assert!(decode("&foo;").is_err());
+        assert!(decode("&foobar;").is_err());
+    }
```
Comment (valkum) [inline instant-xml/src/impls.rs:791]: Let's add some more utf8 values here to be safe.
* `cbdtéda&amp;sü` -> `cbdtéda&sü`
* `cbdtéd&ampü` -> should error I guess

Comment (djc) [reply instant-xml/src/impls.rs:800]: Cool, one of those actually found a bug. Changed `decode()` to work in the byte domain, which simplifies the non-error path code and seems to be slightly faster too.

**Review by valkum — APPROVED** [review]
LGTM. Maybe we want to do some fuzzing in the future.

**Pull request merged**